### PR TITLE
Feature: 'validate'

### PR DESCRIPTION
**Pull request by lsh-0**
* better errors to pass back to bot-lax
* the dry-run option has now become a first class citizen. I think I originally added it for testing purposes?
* better tests for dry-run ('validate') option

this needs to be merged before the corresponding `feat-VALIDATE` of bot-lax can be.

### CODE REVIEW

File: src/publisher/ajson_ingestor.py [hunk 1]
```diff
@@ -144,7 +143,7 @@ def _ingest(data, force=False) -> models.ArticleVersion:
                         'given-version': av.version,
                         'expected-version': last_version.version + 1})
                     LOG.error(msg, extra=log_context)
-                    raise StateError(msg)
+                    raise StateError(codes.PREVIOUS_VERSION_DNE, msg)
```
Comment (giorgiosironi) [inline src/publisher/ajson_ingestor.py]: DNE = Do Not Explain? :)

File: src/publisher/codes.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+PREVIOUS_VERSION_DNE = 'previous-version-does-not-exist'
+idx[PREVIOUS_VERSION_DNE] = "attempt to ingest a version 2 when a version 1 does not exist"
+
+NO_RECORD = 'record-not-found'
+idx[NO_RECORD] = "thrown when we can't find something in the database that we expect to exist"
+
+# ---
+
+def explain(code):
+    global idx
```
Comment (giorgiosironi) [inline src/publisher/codes.py]: is `global` really needed?

File: src/publisher/management/commands/ingest.py [hunk 3]
```diff
@@ -33,28 +34,51 @@ def clean_up(print_queue):
...
+    status = orig_status = lu[action]
+    if dry_run:
+        # this is a dry run, we didn't actually ingest or publish anything
+        status = VALIDATED
+
+    attr = 'datetime_published' if orig_status == PUBLISHED else 'datetime_record_updated'
+    struct = OrderedDict([
+        ('status', status),
+        ('id', log_context['msid']), # good idea? #'id': None if dry_run else av.article.manuscript_id,
+        ('datetime', getattr(av, attr)),
```
Comment (giorgiosironi) [inline src/publisher/management/commands/ingest.py]: `datetime` as a generic word could be interpreted as when this error happened, which can be different from the field. I suggest to add a qualifier like `datetime_record`

File: src/publisher/ajson_ingestor.py (hunk 1 above)
Comment (lsh-0) [inline src/publisher/ajson_ingestor.py]: `DOES NOT EXIST`

File: src/publisher/codes.py (hunk 2 above)
Comment (lsh-0) [inline src/publisher/codes.py]: quite possibly not, probably left over from some other wrangling

File: src/publisher/management/commands/ingest.py (hunk 3 above)
Comment (lsh-0) [inline src/publisher/management/commands/ingest.py]: that value is already a little murky. if the action is PUBLISH, it's the datetime of the publication event, the one that is stored in lax, which is controlled by lax if it's a non v1, but controlled by the xml/production if it's a v1. on ingest it's the time it was created, on validate it's when the job is complete - so, different meanings for different actions but ultimately just a datetime stamp.

**Comment by giorgiosironi** [comment]
The build started failing on `install.sh`?

**Comment by giorgiosironi** [comment]
Possibly related to the other PR for dependencies upgrade leaving around some state difficult to reset in the venv. Worth it to merge the other one and pull it inside this branch.

**Comment by lsh-0** [comment]
> Possibly related to the other PR for dependencies upgrade leaving around some state difficult to reset in the venv.

yeah. so I was getting this on bot-lax and started upgrading the deps but realized I was in lax, so just finished the lax update and returned to bot-lax. 

the pin.sh script is using the feat-VALIDATE branch of bot-lax at the moment while I smooth things out in vagrant.

**Comment by giorgiosironi** [comment]
`develop` has failed consistently on end2end after this merge, and is being rolledback to `approved` after every attempt. There is this error in the logs:
```
ERROR - 2017-07-10 12:38:42,387 - error response -- {"status": "error", "requested-action": "ingest", "status-message": "lax failed attempting to handle our request: message", "datetime": "2017-07-10T12:38:42Z", "token": "eyJzdGF0dXMiOiAidm9yIiwgInJ1biI6ICJmYThjMjMwNy00OGRlLTQ5NWEtYWQyMi1lODAzZDAx\nZmEyYzUiLCAiZXhwYW5kZWRfZm9sZGVyIjogIjgzNzY1ODA1NDQyMjM5MTU4OTMuMS9mYThjMjMw\nNy00OGRlLTQ5NWEtYWQyMi1lODAzZDAxZmEyYzUiLCAidmVyc2lvbiI6ICIxIiwgImZvcmNlIjog\nZmFsc2UsICJlaWZfbG9jYXRpb24iOiAiIn0=\n", "id": "8376580544223915893"}
Traceback (most recent call last):
  File "src/adaptor.py", line 338, in <module>
    bootstrap()
  File "src/adaptor.py", line 334, in bootstrap
    do(*fn())
  File "src/adaptor.py", line 291, in do
    handler(request, outgoing)
  File "/opt/bot-lax-adaptor/venv/local/lib/python2.7/site-packages/newrelic-2.72.1.53/newrelic/api/background_task.py", line 102, in wrapper
    return wrapped(*args, **kwargs)
  File "src/adaptor.py", line 39, in wrap
    result = fn(*args, **kw)
  File "src/adaptor.py", line 243, in handler
    lax_response = call_lax(**params)
  File "src/adaptor.py", line 93, in call_lax
    "message": results['message'],
KeyError: 'message'
```
which seems to point to the lax command not returning a `message` field in its output.

**Comment by lsh-0** [comment]
investigating

**Comment by lsh-0** [comment]
I was mistaken - lax no longer returns a 'message' value on success. it was set to `'message': None` before unless it was a dry run. I instead replaced it with the `'dry-run`: True` (or `False`) but whatever, it would totally break bot-lax without it's `feat-VALIDATE` PR being merged in.

I'm wondering if it's not time to merge bot-lax and lax ... and maybe push the scraping logic into elife-tools.

**Comment by lsh-0** [comment]
https://github.com/elifesciences/lax/pull/241

**Pull request merged**